File: app/src/main/java/com/botijasoftware/utils/Vector2i.java
package com.botijasoftware.utils;

public class Vector2i {
	
	public Vector2i() {}

	public Vector2i(int val) {
		X = val;
		Y = val;
	}
	public Vector2i(int valx, int valy) {
		X = valx;
		Y = valy;
	}
	
	public void setValue(int val) {
		X = val;
		Y = val;
	}
	
	public void setValue(int valx, int valy) {
		X = valx;
		Y = valy;
	}
	
	public void setValue(Vector2i v) {
		X = v.X;
		Y = v.Y;
	}
	
	public Vector2i clone() {
		return new Vector2i(X,Y);
	}
	
	public Vector2i normalize() {
		float len = length();
		if (len != 0.0f) {
			X = (int) (X / len);
			Y = (int) (Y / len);
		}
		return this;
	}
	
	public Vector2i add(Vector2i v) {
		X += v.X;
		Y += v.Y;
		return this;
	}
	
	public Vector2i add(int val) {
		X += val;
		Y += val;
		return this;
	}
	
	public Vector2i sub(Vector2i v) {
		X -= v.X;
		Y -= v.Y;
		return this;
	}
	
	public Vector2i sub(int val) {
		X -= val;
		Y -= val;
		return this;
	}
	
	public Vector2i mul(Vector2i v) {
		X *= v.X;
		Y *= v.Y;
		return this;
	}
	
	public Vector2i mul(int val) {
		X *= val;
		Y *= val;
		return this;
	}
	
	public Vector2i div(Vector2i v) {
		if (v.X != 0.0f && v.Y != 0.0f){
			X = (X / v.X);
			Y = (Y / v.Y);
		}
		return this;
	}
	
	public Vector2i div(int val) {
		if (val != 0.0f){
			float inv = 1.0f/val;
			X = (int) (X * inv);
			Y = (int) (Y * inv);
		}
		return this;
	}
	
	public float lengthsq() {
		return X * X + Y * Y;
	}
	
	public float length() {
		return (float)Math.sqrt( X * X + Y * Y );
	}
	
	public boolean isZero() {
		return (X == 0 && Y == 0);
	}
	
	public boolean equals(int val) {
		return (X == val && Y == val);
	}
	
	public boolean equals(Vector2i v) {
		return (X == v.X && Y == v.Y);
	}
	
	public float distancesq(Vector2i v) {
		return ((X - v.X) * (X - v.X) + (Y - v.Y) * (Y - v.Y));
	}
	
	public float distance(Vector2i v) {
		return (float)Math.sqrt( (X - v.X) * (X - v.X) + (Y - v.Y) * (Y - v.Y) );
	}
	
	public float dot(Vector2i v) {
		return X * v.X + Y * v.Y;
	}
	
	public Vector2i reflect(Vector2i v) {
		//i - (2 * n * dot(i, n))
		sub(v.mul( (int) (2 * dot(v))));
		return this;
	}
	
	public Vector2i lerp(Vector2i v, float factor) {
		if (factor <=0.0f) {
			return this;
		}
		else if (factor >= 1.0f) {
			X = v.X;
			Y = v.Y;
			return this;
		}
		
		X = X + (int)(factor * ( v.X - X));
		Y = Y + (int)(factor * ( v.Y - Y));
		
		return this;
	}
	
	public Vector2i zero() {
		X = 0;
		Y = 0;
		return this;
	}
	
	
	public static Vector2i lerp(Vector2i v, Vector2i u, float factor) {
		if (factor <=0.0f) {
			return v.clone();
		}
		else if (factor >= 1.0f) {
			return u.clone();
		}
		
		int x = v.X + (int)(factor * ( u.X - v.X));
		int y = v.Y + (int)(factor * ( u.Y - v.Y));
		
		return new Vector2i(x,y);
	}
	
	public int X = 0;
	public int Y = 0;
	public static final Vector2i ZERO = new Vector2i(0, 0);
        public static final Vector2i ONE = new Vector2i(1, 1);
        public static final Vector2i RIGHT = new Vector2i(1, 0);
        public static final Vector2i UP = new Vector2i(0, 1);

}

File: app/src/main/java/com/botijasoftware/utils/Vector2i.kt
package com.botijasoftware.utils

class Vector2i {

    constructor() {
        X = 0
        Y = 0
    }

    constructor(value: Int) {
        X = value
        Y = value
    }

    constructor(valx: Int, valy: Int) {
        X = valx
        Y = valy
    }

    fun setValue(value: Int) {
        X = value
        Y = value
    }

    fun setValue(valx: Int, valy: Int) {
        X = valx
        Y = valy
    }

    fun setValue(v: Vector2i) {
        X = v.X
        Y = v.Y
    }

    fun clone(): Vector2i {
        return Vector2i(X, Y)
    }

    fun normalize(): Vector2i {
        val len = length()
        if (len != 0.0f) {
            X = (X / len).toInt()
            Y = (Y / len).toInt()
        }
        return this
    }

    fun add(v: Vector2i): Vector2i {
        X += v.X
        Y += v.Y
        return this
    }

    fun add(value: Int): Vector2i {
        X += value
        Y += value
        return this
    }

    fun sub(v: Vector2i): Vector2i {
        X -= v.X
        Y -= v.Y
        return this
    }

    fun sub(value: Int): Vector2i {
        X -= value
        Y -= value
        return this
    }

    fun mul(v: Vector2i): Vector2i {
        X *= v.X
        Y *= v.Y
        return this
    }

    fun mul(value: Int): Vector2i {
        X *= value
        Y *= value
        return this
    }

    operator fun div(v: Vector2i): Vector2i {
        if (v.X != 0 && v.Y != 0) {
            X /= v.X
            Y /= v.Y
        }
        return this
    }

    operator fun div(value: Int): Vector2i {
        if (value != 0) {
            val inv = 1.0f / value
            X = (X * inv).toInt()
            Y = (Y * inv).toInt()
        }
        return this
    }

    fun lengthsq(): Float {
        return (X * X + Y * Y).toFloat()
    }

    fun length(): Float {
        return Math.sqrt((X * X + Y * Y).toDouble()).toFloat()
    }

    val isZero: Boolean
        get() = X == 0 && Y == 0

    fun equals(value: Int): Boolean {
        return X == value && Y == value
    }

    fun equals(v: Vector2i): Boolean {
        return X == v.X && Y == v.Y
    }

    fun distancesq(v: Vector2i): Float {
        return ((X - v.X) * (X - v.X) + (Y - v.Y) * (Y - v.Y)).toFloat()
    }

    fun distance(v: Vector2i): Float {
        return Math.sqrt(((X - v.X) * (X - v.X) + (Y - v.Y) * (Y - v.Y)).toDouble()).toFloat()
    }

    fun dot(v: Vector2i): Float {
        return (X * v.X + Y * v.Y).toFloat()
    }

    fun reflect(v: Vector2i): Vector2i {
        //i - (2 * n * dot(i, n))
        sub(v.mul((2 * dot(v)).toInt()))
        return this
    }

    fun lerp(v: Vector2i, factor: Float): Vector2i {
        if (factor <= 0.0f) {
            return this
        } else if (factor >= 1.0f) {
            X = v.X
            Y = v.Y
            return this
        }

        X += (factor * (v.X - X)).toInt()
        Y += (factor * (v.Y - Y)).toInt()

        return this
    }

    fun zero(): Vector2i {
        X = 0
        Y = 0
        return this
    }

    var X: Int = 0
    var Y: Int = 0

    companion object {


        fun lerp(v: Vector2i, u: Vector2i, factor: Float): Vector2i {
            if (factor <= 0.0f) {
                return v.clone()
            } else if (factor >= 1.0f) {
                return u.clone()
            }

            val x = v.X + (factor * (u.X - v.X)).toInt()
            val y = v.Y + (factor * (u.Y - v.Y)).toInt()

            return Vector2i(x, y)
        }

        val ZERO = Vector2i(0, 0)
        val ONE = Vector2i(1, 1)
        val RIGHT = Vector2i(1, 0)
        val UP = Vector2i(0, 1)
    }

}
